### PR TITLE
fix: error when uploading same file again

### PR DESCRIPTION
**Pull request by forouher**
Reasoning: Temp files aren't deleted after upload.
If the user tries to upload the same file twice,
the temp file cannot be created again, causing an

    throw new RuntimeException("could not create temporary file");

We can't delete or overwrite the existing temp file, as the upload
might still be in progress.

There are other ways to fix this issue, but this is the simplest solution,
as far as I can see.

### CODE REVIEW

File: src/com/seafile/seadroid2/ui/activity/BrowserActivity.java [hunk 1]
```diff
@@ -1047,7 +1047,7 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
 
                 try {
                     if (!tempDir.exists()) {
```
Comment (Logan676) [inline src/com/seafile/seadroid2/ui/activity/BrowserActivity.java]: if you changed to 
```
File tempDir = new File(DataManager.getExternalTempDirectory(), "saf_temp" + "/" + "upload-"+System.currentTimeMillis());
```
the `tempDir.exists()` will always be false, and should be removed.

**Comment by Logan676** [comment]
Maybe it\`s stupid to ask, but I wonder if we can reuse the downloaded file, and why?
IMHO, only when the downloaded file was modified (remote not consistent with local), then we need to create a new one to substitute the old (say update), but is that being the common case?

**Comment by forouher** [comment]
> Maybe it`s stupid to ask, but I wonder if we can reuse the downloaded file, and why?

Scenario:
1. The phone has no network connection
2.. The user tries to upload file "folderA/filename.jpg"
3. The user tries to upload file "folderB/filename.jpg"
4. the phone has network connection again
5. the user restarts the failed uploads from the menu
6. The wrong files are being uploaded.

**Comment by forouher** [comment]
regarding your comments in PR #308:

> BTW, I have tested this patch, turned out it didn`t fix the "unable to upload, no path available" errors.
> In order to reproduce it,
> Choose upload menu to open the document provider
> Choose Images menu to open the gallery like gridview
> Choose a gallery to open the photo list
> Choose a photo or long press the list to choose multiple photos
> the message "unable to upload, no path available" appears

Did you have this PR (#307) merged when testing this?



> Do I have to merge them into 287 in order to "assemble" the whole feature for testing.

- #307 is necessary to fix some of the "no path available" issues.
- #287 is not necessary. And if you guys want to release soon, I would suggest to delay #287 until the next release. #287 doesn't touch any of the upload issues we've been discussing during the last days. It only modifies the *provider-side* of the SAF. Which is only used when accessing the SAF from other apps (Gmail, Openkeychain,...)
- With #308 it was my intention to avoid having stale images in the image gallery. The problem is that the Android Gallery doesn't refresh it's image list when Seafile clears its cache. So after you've purged your Seafile cache, you would be able to select non-existent files. Doing so will cause a "no path available" error mentioned.

I'll try to find another way to notify the Android Gallery when Seafile deletes an image in it's cache.



So in summary:

After applying this PR on master, the only remaining issue should be:
- trying to upload an image/video that was lying in the Seafile cache, but has since been removed by Seafile cache purging.

If you notice any other issues (e.g. when uploading files from the "Downloads" folder), please try to get a Log excerpt, especially the exception mentioned in "Could not open requested document"

**Comment by forouher** [comment]
This should fix all remaining issues regarding file upload.

**Comment by Logan676** [comment]
> Scenario:
> 1. The phone has no network connection
> 2.. The user tries to upload file "folderA/filename.jpg"
> 3. The user tries to upload file "folderB/filename.jpg"
> 4. the phone has network connection again
> 5. the user restarts the failed uploads from the menu
> 6. The wrong files are being uploaded.  

No, If the phone has no network connection, we should stop user from uploading anything. So the expected scenario should be like this
* 1. The phone has no network connection
* 2. Stop the user from trying to upload file "folderA/filename.jpg"
* 3. Stop the user from trying to upload file "folderB/filename.jpg"
* 4. The phone has network connection again
* 5. The user restarts the failed uploads from the menu
* 6. The files are being uploaded.

> If you notice any other issues (e.g. when uploading files from the "Downloads" folder), please try to get a Log excerpt, especially the exception mentioned in "Could not open requested document"

Now it works fine when uploading files from `Images` or `Videos` folder. But when uploading files from `Downloads` folder, it failed sometimes, the error log is 
```
 java.io.FileNotFoundException: No such file or directory
            at android.database.DatabaseUtils.readExceptionWithFileNotFoundExceptionFromParcel(DatabaseUtils.java:146)
            at android.content.ContentProviderProxy.openTypedAssetFile(ContentProviderNative.java:682)
            at android.content.ContentResolver.openTypedAssetFileDescriptor(ContentResolver.java:1067)
            at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:907)
            at android.content.ContentResolver.openInputStream(ContentResolver.java:632)
            at com.seafile.seadroid2.ui.activity.BrowserActivity$SAFLoadRemoteFileTask.doInBackground(BrowserActivity.java:1055)
            at com.seafile.seadroid2.ui.activity.BrowserActivity$SAFLoadRemoteFileTask.doInBackground(BrowserActivity.java:1031)
            at android.os.AsyncTask$2.call(AsyncTask.java:288)
            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)
```

**Comment by forouher** [comment]
Hi Logan,

> Now it works fine when uploading files from Images or Videos folder. But when uploading files from 
> Downloads folder, it failed sometimes, the error log is 

For for files where this fails, could you please:

1. have a look at the debug message preceding the exception:
--> Log.d(DEBUG_TAG, "Uploading file from uri: " + uri);
and make sure that the uri matches the file you've selected

2. verify with a 3rd party file manager (eg. "ES File Explorer") that the file in question actually exists.

thanks
Dariush

**Comment by Logan676** [comment]
> verify with a 3rd party file manager (eg. "ES File Explorer") that the file in question actually exists  

The file wasn\`t exists any more

**Comment by forouher** [comment]
> The file wasn`t exists any more

Thanks. Then this is not a Seafile issue.

**Comment by forouher** [comment]
> No, If the phone has no network connection, we should stop user from uploading anything. So the 
> expected scenario should be like this

Okay, but currently that's not the case. And enforcing in a robust way would not be trivial.

But ignoring that, overwriting the files would still be an unsafe think to do, in my opinion. After creating the temp file, the uploading itself is performed by the UploadTask asynchronously. The BrowserActivity doesn't know when the UploadTask will be finished. A 100MB upload via 3G might take many minutes. If the user uploads the second file before the first is finished, there might be data corruption.
